File: 13.Linked_List/python/6.DeletingMiddleNode.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

# Function to build a linked list from a list
def build_linked_list(arr):
    if not arr:
        return None

    head = Node(arr[0])
    current = head

    for val in arr[1:]:
        current.next = Node(val)
        current = current.next

    return head

# Function to remove the middle node
def remove_middle_node(head, n):
    if n == 1:
        return None  # Only one node

    middle = n // 2  # 0-based middle index

    if middle == 0:
        return head.next  # Remove the first node in 2-element list

    current = head
    for _ in range(middle - 1):
        current = current.next

    if current.next:
        current.next = current.next.next

    return head

# Function to print the linked list
def print_list(head):
    current = head
    while current:
        print(current.data, end=" ")
        current = current.next
    print()

# ----------- INPUT -----------
inputs = [3, 1, 3, 4, 5, 5, 2]
n = len(inputs)

# ----------- BUILD, PROCESS, PRINT -----------
head = build_linked_list(inputs)
new_head = remove_middle_node(head, n)
print_list(new_head)  # Output: 3 1 3 5 5 2

File: 13.Linked_List/javascript/6.DeletingMiddleNode.js
// Node class definition
class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

// Function to build a linked list from an array
function buildLinkedList(arr) {
    if (arr.length === 0) return null;

    let head = new Node(arr[0]);
    let current = head;

    for (let i = 1; i < arr.length; i++) {
        current.next = new Node(arr[i]);
        current = current.next;
    }

    return head;
}

// Function to remove the middle node
function removeMiddleNode(head, n) {
    if (n === 1) return null; // Only one node, remove it

    let middle = Math.floor(n / 2); // 0-based index

    if (middle === 0) return head.next; // If only two nodes

    let current = head;
    for (let i = 0; i < middle - 1; i++) {
        current = current.next;
    }

    if (current.next) {
        current.next = current.next.next;
    }

    return head;
}

// Function to print linked list
function printList(head) {
    let current = head;
    let result = [];
    while (current) {
        result.push(current.data);
        current = current.next;
    }
    console.log(result.join(" "));
}

// ----------- INPUT -----------
const inputs = [3, 1, 3, 4, 5, 5, 2];
const n = inputs.length;

// ----------- BUILD, PROCESS, PRINT -----------
let head = buildLinkedList(inputs);
let newHead = removeMiddleNode(head, n);
printList(newHead);  // Output: 3 1 3 5 5 2
